### PR TITLE
[BEAM-2958] Adding a top-level user agent string to PipelineOptions

### PR DESCRIPTION
**Pull request by youngoli**
Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Make sure there is a [JIRA issue](https://issues.apache.org/jira/projects/BEAM/issues/) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [x] Each commit in the pull request should have a meaningful subject line and body.
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue.
 - [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [x] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---

Adding a string to PipelineOptions that can be used for setting a user agent string with information about the distribution of Beam being used. That string can be sent to external services too.

This PR is probably not completely done yet. While the actual user agent string is in and working, more might have to be done to replace BigTableIO and DataflowRunner usages of user agents to use this new one. I'm still working on that right now.

### CODE REVIEW

**Comment by youngoli** [comment]
R: @lukecwik

**Comment by lukecwik** [comment]
retest this please

File: sdks/java/core/src/main/java/org/apache/beam/sdk/options/PipelineOptions.java [hunk 1]
```diff
@@ -353,4 +354,30 @@ public Long create(PipelineOptions options) {
...
   }
+
+  /**
+   * A user agent string describing the pipeline to external services.
+   *
+   * <p>The string defaults to "[name]/[version]" based on the properties of the Beam release.
+   */
+  @Description("A user agent string describing the pipeline to external services. "
+      + "The default name is \"[name]/[version]\""
+      + " where name and version are properties of the Beam release.")
```
Comment (lukecwik) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/options/PipelineOptions.java]: `of the Beam release.` -> `of the Apache Beam release.`

File: sdks/java/core/src/main/java/org/apache/beam/sdk/options/PipelineOptions.java [hunk 2]
```diff
@@ -353,4 +354,30 @@ public Long create(PipelineOptions options) {
...
+  @Default.InstanceFactory(UserAgentFactory.class)
+  String getUserAgent();
+  void setUserAgent(String userAgent);
+
+  /**
+   * Returns a user agent string constructed from {@link ReleaseInfo#getName()} and
+   * {@link ReleaseInfo#getVersion()}, in the format "[name]/[version]".
```
Comment (lukecwik) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/options/PipelineOptions.java]: `"[name]/[version]"` -> `{@code [name]/[version]}`

File: sdks/java/core/src/main/java/org/apache/beam/sdk/options/PipelineOptions.java [hunk 3]
```diff
@@ -353,4 +354,30 @@ public Long create(PipelineOptions options) {
...
+   */
+  class UserAgentFactory implements DefaultValueFactory<String> {
+    @Override
+    public String create(PipelineOptions options) {
+      ReleaseInfo info = ReleaseInfo.getReleaseInfo();
+      return
```
Comment (lukecwik) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/options/PipelineOptions.java]: nit: whitespace around return looks awkward

File: sdks/java/core/src/main/java/org/apache/beam/sdk/options/PipelineOptions.java [hunk 4]
```diff
@@ -353,4 +354,28 @@ public Long create(PipelineOptions options) {
...
+
+  /**
+   * Returns a user agent string constructed from {@link ReleaseInfo#getName()} and
+   * {@link ReleaseInfo#getVersion()}, in the format "[name]/[version]".
+   */
+  class UserAgentFactory implements DefaultValueFactory<String> {
+    @Override
+    public String create(PipelineOptions options) {
+      ReleaseInfo info = ReleaseInfo.getReleaseInfo();
+      return String.format("%s/%s", info.getName(), info.getVersion()).replace(" ", "_");
```
Comment (lukecwik) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/options/PipelineOptions.java]: Is there a reason why we replace `" "` with `"_"`?

Comment (youngoli) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/options/PipelineOptions.java]: That's how BigtableIO originally had it when reading from ReleaseInfo. Why it did that replacement I'm not sure.

Comment (lukecwik) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/options/PipelineOptions.java]: It turns out they are following https://www.ietf.org/rfc/rfc2616.txt where:
```
user agent         = 1*(product | comment)
product            = token ["/" product-version]
product-version    = token
```

Add to the user agent comment that this field should represent a user agent as per RFC2616 and give the BNF form that I describe above. A token is limited to any characters not including a separator.

**Comment by youngoli** [comment]
retest this please

File: sdks/java/core/src/main/java/org/apache/beam/sdk/options/PipelineOptions.java [hunk 5]
```diff
@@ -353,4 +354,38 @@ public Long create(PipelineOptions options) {
...
+   *
+   * <p>https://www.ietf.org/rfc/rfc2616.txt
+   *
+   * <p>It should follow the BNF Form:
+   * <pre><code>
+   * user agent         = 1*(product | comment)
+   * product            = token ["/" product-version]
+   * product-version    = token
+   * </code></pre>
+   * Where a token is a series of without a separator.
```
Comment (lukecwik) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/options/PipelineOptions.java]: Awkward sentence: `Where a token is a series of without a separator.`

File: sdks/java/core/src/main/java/org/apache/beam/sdk/options/PipelineOptions.java [hunk 6]
```diff
@@ -353,4 +354,38 @@ public Long create(PipelineOptions options) {
...
+   *
+   * <p>The string defaults to {@code [name]/[version]} based on the properties of the Beam release.
```
Comment (lukecwik) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/options/PipelineOptions.java]: `Beam` -> `Apache Beam`

File: sdks/java/core/src/main/java/org/apache/beam/sdk/options/PipelineOptions.java (hunk 2 above)
Comment (lukecwik) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/options/PipelineOptions.java]: ?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/options/PipelineOptions.java (hunk 5 above)
Comment (youngoli) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/options/PipelineOptions.java]: Whoops, forgot a word.

**Comment by lukecwik** [comment]
Failed in known Gearpump test issue. LGTM, merging

**Pull request closed without merging**